lib/ex_doc/retriever.ex

```diff
@@ -370,36 +318,31 @@ defmodule ExDoc.Retriever do
     {:docs_v1, _, _, _, _, _, docs} = module_data.docs
     optional_callbacks = module_data.name.behaviour_info(:optional_callbacks)
 
-    for {{kind, _, _}, _, _, _, _} = doc <- docs, kind in [:callback, :macrocallback] do
+    for {{kind, _, _}, _, _, _, _} = doc <- docs, kind in module_data.callback_types do
       get_callback(doc, source, optional_callbacks, module_data)
     end
   end
 
   defp get_callbacks(_, _), do: []
 
   defp get_callback(callback, source, optional_callbacks, module_data) do
+    callback_data = module_data.language.callback_data(callback, module_data)
+
     {:docs_v1, _, _, content_type, _, _, _} = module_data.docs
     {{kind, name, arity}, anno, signature, doc, metadata} = callback
-    actual_def = actual_def(name, arity, kind)
+    actual_def = callback_data.actual_def
     doc_line = anno_line(anno)
     signature = signature(signature)
+    specs = callback_data.specs
 
-    {specs, line, signature} =
-      case Map.fetch(module_data.callbacks, actual_def) do
-        {:ok, specs} ->
-          {:type, anno, _, _} = hd(specs)
-          line = anno_line(anno)
-
-          specs = Enum.map(specs, &Code.Typespec.spec_to_quoted(name, &1))
-          signature = signature || get_typespec_signature(hd(specs), arity)
-          {specs, line, signature}
-
-        :error ->
-          {[], doc_line, signature || "#{name}/#{arity}"}
-      end
+    signature =
+      signature ||
+        (callback_data.signature_fallback && callback_data.signature_fallback.()) ||
```

Comment:
Make it default to a function that returns nil so you can do:

```suggestion
        callback_data.signature_fallback.() ||
```

Reply:
Good call